rfd/0036-github-actions.md

```diff
@@ -0,0 +1,246 @@
+
+
+
+### Vulnerabilities 
+
+#### [Bypassing a Required Reviewer](https://medium.com/cider-sec/bypassing-required-reviews-using-github-actions-6e1b29135cc7)
+
+The Github actions token is granted write permissions to pull requests and with this, an attacker can run a command that approves the pull request with malicious code. An attacker could only merge this pull request if the repository settings didn't require 2+ reviewers or if `CODEOWNERS` wasn't exhaustively set. The repo this bot manages has both of these attributes. `CODEOWNERS` cover all paths of the repository and each pull request requires 3 reviewers before it can be merged. 
+
+#### User Obtains the Github Token
+
+The maximum permissions on any given workflow are `pull-requests:write` and `actions:write`. If an attacker were to somehow obtain the Github token, the following is what they would have access to: 
+
+- [Permissions](https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-actions) on Actions
+- [Permissions](https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-pull-requests) on Pull Requests
+
+Concerning actions an attacker can perform:  
+
+| Action     | Scenario/Mitigation  |
+| ----------- | ----------- |
+| Cancels a run     |   An attacker could get all approvals on a pull request, trigger a `pull_request_target` event (such as `synchronize` which is a pushed commit to PR), and cancel a run in that commit with malicious code. If only internal contributors/CODEOWNERS have the ability to merge a PR from a fork, the security in place should be enough.     |
+| Delete/edit comments on issues or pull requests.   | An attacker could edit or delete important comments. There doesn't seem to be a way to get a malicious commit in master this way.  |
+| Delete logs of a workflow. | An attacker could delete a logs of a workflow that would otherwise fail in their favor. This alone wouldn't be enough to get a malicious commit in for external contributors if the internal contributor that merges their pull request in checks that `Assign` and `Check` workflows pass. Because we do not invalidate reviews for internal contributors when a new commit is pushed and they have the ability to merge their own pull request in, it could be possible to get a malicious commit in. |
+| Re-run a workflow.  | An attacker could re-run a workflow though there wouldn't be any benefit to them even if code was changes. Workflow would just run against the new code and pass/fail accordingly. | 
+| Update an issue or a pull request. | An attacker could update the contents of a pull request or issue.  There doesn't seem to be a way to get a malicious commit in master this way.| 
```

Comment:
Could an attacker change the contents of any PR?

For example, could an attacker change the contents of an internal PR right before it is merged?

Reply:
Correct, but not the actual code, just the description. If you're referring to the actual code, they would need to create a commit and push it to the pull request before the merge. The user would need `contents:write` ( see https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-contents) to post a commit, but we set permissions to none for contents.